### PR TITLE
Implement azureblobstorage target in luisy

### PR DESCRIPTION
**Pull request by zyko**
Small rework of @hueb1337 first draft.
This is most likely no final version, but might be a good start for a discussion.
Maybe we can discuss a few things synchronously @windisch 

1. Are we using `outdir` in CloudTargets?
2. Are we fine with fixed pickling objects? Or do we want to have it flexible?
3. (most important) Do we currently have access to some blob storage so that we can try the implementation online?

### CODE REVIEW

File: luisy/targets.py [hunk 1]
```diff
@@ -242,6 +245,75 @@ def read(self):
...
+    def write(self, obj):
+        """
+        Write object to Azure Blob Storage
+        Args:
+            obj (object): object that can be serialized using `pickle` and is to be written to
+                the blob output
+        """
+
+        try:
+            serialized_obj = pickle.dumps(obj)
```
Comment (windisch) [inline luisy/targets.py]: as discussed synchronically: We may restrict to the case that `obj` is a pyspark-frame and save to parquet files. Then this target would work the same way as a `DeltaTableTarget`.

**Review by windisch — APPROVED** [review]

**Comment by zyko** [comment]
@windisch what do you think about the latest changes of [this](https://github.com/zyko/luisy/pull/4/commits/bd52643fde93cdcc881129361f0f1aef46949c9c) commit ?
Unfortunately, I couldn't find a better way of removing a file. But I'm also not sure if we need this. Would it be possible just always to overwrite?

File: luisy/decorators.py [hunk 2]
```diff
@@ -258,4 +256,29 @@ def output_decorator(cls):
...
+        cls.target_cls = AzureBlobStorageTarget
+        cls.target_kwargs = {
+            'blob': blob,
+        }
+        return cls
+
+    return decorator
+
+
+def azure_blob_storage_input(blob=None):
```
Comment (windisch) [inline luisy/decorators.py]: Should we explicitly state and forward all constructor arguments of the AzureBlobStorageTarget?

File: luisy/targets.py [hunk 3]
```diff
@@ -242,6 +250,64 @@ def read(self):
...
+        self.container = container
+        self.account = account
+        self.abfss_path = abfss_path
+        self.inferschema = inferschema
+
+    def make_dir(self, path):
+        pass
+
+    def remove(self):
+        from py4j.java_gateway import java_import
```
Comment (windisch) [inline luisy/targets.py]: Probably code that will break in future ans we definitely have to touch (due to all the private methods), may be a bit risky that we delete data accidentally. 

As we always overwrite, we could also leave this function void for a start?

File: luisy/targets.py [hunk 4]
```diff
@@ -242,6 +250,64 @@ def read(self):
...
+        return (f"abfss://{self.container}@{self.account}.dfs.core.windows.net/"
+                f"{self.abfss_path}.{self.file_ending}")
+
+    def exists(self):
+        """
+        Checks whether the file exists in Azure Blob Storage
+
+        """
+        try:
+            self.spark.read.format(self.file_ending).load(self.path).limit(1).count()
```
Comment (windisch) [inline luisy/targets.py]: Good idea!

File: luisy/targets.py [hunk 5]
```diff
@@ -242,6 +250,64 @@ def read(self):
...
+        except Exception:
+            return False
+
+    def write(self, df):
+        """
+        Write Pyspark DataFrame to Azure Blob Storage
+        Args:
+            df (pyspark.DataFrame): DataFrame that is to be stored in Azure Blob Storage
+        """
+        df.write.format(self.file_ending).save(self.path)
```
Comment (windisch) [inline luisy/targets.py]: Should we overwrite here?

**Review by windisch — APPROVED** [review]

File: luisy/targets.py (hunk 3 above)
Comment (windisch) [reply luisy/targets.py]: Decided to leave it out and to not remove for a start

**Pull request merged**